Compile/src/Producers/LeptonSFProducer.cc

```diff
@@ -35,15 +38,133 @@ std::string LeptonSFProducer::GetProducerId() const { return "LeptonSFProducer";
+            float sf1u = 1 / this->GetScaleFactor(-1, *lep1);
+            float sf1d = 1 / this->GetScaleFactor(+1, *lep1);
+            float sf2u = 1 / this->GetScaleFactor(-1, *lep2);
+            float sf2d = 1 / this->GetScaleFactor(+1, *lep2);
+            product.m_optionalWeights["zl1" + weightName + "Up"] = sf1u;
+            product.m_optionalWeights["zl1" + weightName + "Down"] = sf1d;
+            product.m_optionalWeights["zl2" + weightName + "Up"] = sf2u;
+            product.m_optionalWeights["zl2" + weightName + "Down"] = sf2d;
+            // proper error propagation needed
+            product.m_optionalWeights["lepton" + weightName + "Up"] = sf1u * sf2u;
+            product.m_optionalWeights["lepton" + weightName + "Down"] = sf1d * sf2d;
+            if (settings.GetDebugVerbosity() > 1) {
+                LOG(DEBUG) << "Lepton1 SF up: " << sf1u << ", Lepton1 SF down: " << sf1d;
+                LOG(DEBUG) << "Lepton2 SF up: " << sf2u << ", Lepton2 SF down: " << sf2d;
+                LOG(DEBUG) << "Total SF up: " << sf1u * sf2u << ", Total SF down: " << sf1d * sf2d;
+            }
+        }
+    } else {
+        LOG(DEBUG) << "No valid Z found. Only applying SFs to Z leptons.";
+    }
+}
+
+void LeptonSFProducer::SetEtaAxis2D(std::string histoname)
+{
+    // map with eta on x, pt on y is default!
+    if (histoname.find("abseta") != std::string::npos) {
+        m_absoluteeta = true;
+        if (histoname.find("abseta_pt") != std::string::npos) {
+            m_reversed_axes = false;
+            LOG(INFO) << this->GetProducerId() << ": abs(eta)-pt map used for SF";
+        } else if (histoname.find("pt_abseta") != std::string::npos) {
+            LOG(INFO) << "pt-abs(eta) map (reversed axes) used for SF";
+            m_reversed_axes = true;
+        } else {
+            LOG(FATAL) << this->GetProducerId() << ": Neither pt-abs(eta) nor abs(eta)-pt map found for SF.";
+        }
+    }
+    // Attention: abseta needs to be caught in advance!
+    else if (histoname.find("eta") != std::string::npos) {
+        m_absoluteeta = false;
+        if (histoname.find("eta_pt") != std::string::npos) {
+            m_reversed_axes = false;
+            LOG(INFO) << this->GetProducerId() << ": eta-pt map used for SF";
+        } else if (histoname.find("pt_eta") != std::string::npos) {
+            m_reversed_axes = true;
+            LOG(INFO) << this->GetProducerId() << ": pt-eta map (reversed axes) used for SF";
+        } else {
+            LOG(FATAL) << this->GetProducerId() << ": Neither pt-eta nor eta-pt map found for SF.";
+        }
+    } else
+        LOG(FATAL) << this->GetProducerId() << ": No known abs(eta)- or eta-pt map found for SF.";
+    return;
+}
+
+void LeptonSFProducer::SetOverflowPtBin()
+{
+    int nEta = m_reversed_axes ? sfhisto->GetNbinsY() : sfhisto->GetNbinsX();
+    int nPt = m_reversed_axes ? sfhisto->GetNbinsX() : sfhisto->GetNbinsY();
+    for (int i = 1; i <= nEta; ++i) {
```

Comment:
Maybe add a comment for clarity, why it starts with 1 (underflow).

Reply:
Ok